pkg/image/prune/prune.go

```diff
@@ -240,28 +240,41 @@ func NewPruner(options PrunerOptions) (Pruner, error) {
 	}
 	algorithm.namespace = options.Namespace
 
-	g := graph.New()
-	addImagesToGraph(g, options.Images, algorithm)
-	addImageStreamsToGraph(g, options.Streams, options.LimitRanges, algorithm)
-	addPodsToGraph(g, options.Pods, algorithm)
-	addReplicationControllersToGraph(g, options.RCs)
-	if err := addBuildConfigsToGraph(g, options.BCs); err != nil {
-		return nil, err
+	p := &pruner{
+		algorithm:      algorithm,
+		registryClient: options.RegistryClient,
+		registryURL:    options.RegistryURL,
 	}
-	if err := addBuildsToGraph(g, options.Builds); err != nil {
+
+	if err := p.buildGraph(options); err != nil {
 		return nil, err
 	}
-	addDaemonSetsToGraph(g, options.DSs)
-	addDeploymentsToGraph(g, options.Deployments)
-	addDeploymentConfigsToGraph(g, options.DCs)
-	addReplicaSetsToGraph(g, options.RSs)
 
-	return &pruner{
-		g:              g,
-		algorithm:      algorithm,
-		registryClient: options.RegistryClient,
-		registryURL:    options.RegistryURL,
-	}, nil
+	return p, nil
+}
+
+// buildGraph builds a graph
+func (p *pruner) buildGraph(options PrunerOptions) error {
+	p.g = graph.New()
+
+	p.addImagesToGraph(options.Images, p.algorithm)
+	p.addImageStreamsToGraph(options.Streams, options.LimitRanges, p.algorithm)
+	p.addPodsToGraph(options.Pods, p.algorithm)
+	p.addReplicationControllersToGraph(options.RCs)
+	if err := p.addBuildConfigsToGraph(options.BCs); err != nil {
+		// TODO: either make the other methods error out on image reference parsing or ignore this error
+		return err
+	}
+	if err := p.addBuildsToGraph(options.Builds); err != nil {
+		// TODO: either make the other methods error out on image reference parsing or ignore this error
```

Comment:
@legionus what do you think?

Reply:
i'd prefer we report a warning if we can't parse an image reference, but don't prevent pruning.  I think we had this conversation before, but my view is that if you don't have valid reference, we shouldn't care what you were trying to reference.

Reply:
> i'd prefer we report a warning if we can't parse an image reference, but don't prevent pruning. I think we had this conversation before, but my view is that if you don't have valid reference, we shouldn't care what you were trying to reference.

I agree. Changed to warning.

Reply:
I disagree.

There can be two cases:

1. Docker change their policy on references (for example, they may allow the `+` symbol in tags). And we can change our policy on tags as well. In this case the old version of `oc` will fail to parse this reference, and it will happily remove images with such references ("graceful fail", yep).

2. Someone was able to create a resource with an invalid image reference. This state is bad enough to start investigations. We shouldn't ignore bugs in our code.

Reply:
if it's not a reference we can parse, then there can be no pods using it, by definition, no?

Whether that's a bug in our code or not doesn't really matter, if we can't parse it, it's not a reference.

And while i agree that the ability to create invalid references is bad(if it were to happen), the fact is the code paths are different (validation vs resolution) and pruning is not the place to reconcile/debug that.

Reply:
Pruning is a client-side operation and it may have an outdated parser.

Reply:
> Whether that's a bug in our code or not doesn't really matter, if we can't parse it, it's not a reference.

@bparees But if you can't parse reference then you don't add a dependency and you can remove what is used right now.

Is not that a problem?

Reply:
@miminar It's NAK for me.